src/client/client.rs

```diff
@@ -268,70 +98,27 @@ where
         qos: QualityOfService,
         retain: bool,
     ) -> Result<(), ReasonCode> {
-        match self.config.mqtt_version {
-            MqttVersion::MQTTv3 => Err(ReasonCode::UnsupportedProtocolVersion),
-            MqttVersion::MQTTv5 => self.send_message_v5(topic_name, message, qos, retain).await,
-        }
-    }
+        let identifier = self
+            .raw
+            .send_message(topic_name, message, qos, retain)
+            .await?;
 
-    async fn subscribe_to_topics_v5<'b, const TOPICS: usize>(
-        &'b mut self,
-        topic_names: &'b Vec<&'b str, TOPICS>,
-    ) -> Result<(), ReasonCode> {
-        if self.connection.is_none() {
-            return Err(ReasonCode::NetworkError);
-        }
-        let conn = self.connection.as_mut().unwrap();
-        let len = {
-            let mut subs = SubscriptionPacket::<'b, TOPICS, MAX_PROPERTIES>::new();
-            let mut i = 0;
-            loop {
-                if i == TOPICS {
-                    break;
+        // QoS1
+        if qos == QoS1 {
+            match self.raw.poll::<0>().await? {
+                Event::Puback(ack_identifier) => {
+                    if identifier == ack_identifier {
+                        Ok(())
+                    } else {
+                        Err(ReasonCode::PacketIdentifierNotFound)
+                    }
                 }
-                subs.add_new_filter(topic_names.get(i).unwrap(), self.config.max_subscribe_qos);
-                i = i + 1;
-            }
-            subs.encode(self.buffer, self.buffer_len)
-        };
-
-        if let Err(err) = len {
-            error!("[DECODE ERR]: {}", err);
-            return Err(ReasonCode::BuffError);
-        }
-
-        conn.send(&self.buffer[0..len.unwrap()]).await?;
-
-        let reason: Result<Vec<u8, TOPICS>, BufferError> = {
-            let read =
-                { receive_packet(self.buffer, self.buffer_len, self.recv_buffer, conn).await? };
-
-            let mut packet = SubackPacket::<'b, TOPICS, MAX_PROPERTIES>::new();
-            if let Err(err) = packet.decode(&mut BuffReader::new(self.buffer, read)) {
-                Err(err)
-            } else {
-                Ok(packet.reason_codes)
-            }
-        };
-
-        if let Err(err) = reason {
-            error!("[DECODE ERR]: {}", err);
-            return Err(ReasonCode::BuffError);
-        }
-        let reasons = reason.unwrap();
-        let mut i = 0;
-        loop {
-            if i == TOPICS {
-                break;
+                Event::Disconnect(reason) => Err(reason),
+                _ => Err(ReasonCode::ImplementationSpecificError),
```

Comment:
Not sure how to solve it, but this will result in the loss of data when at this moment the raw client decodes a Publish packet. Let's discuss it in person.

Reply:
Seeing how it is fixed in our codebase, let's merge it like it is now. Can you please add comment about the potential data loss somewhere so it isn't forgotten?

Reply:
Ok, I will add a comment. I think this is the same behavior as before (only returning a different error when that happens). In this PR, I didn't fix `MqttClient` struct, I only provide `RawMqttClient` which enables lower-level handling and thus mitigating this issue.